src/spark_rapids_ml/knn.py

```diff
@@ -493,3 +521,100 @@ def _get_cuml_transform_func(
         raise NotImplementedError(
             "'_CumlModel._get_cuml_transform_func' method is not implemented. Use 'kneighbors' instead."
         )
+
+    def exactNearestNeighborsJoin(
+        self,
+        datasetA: DataFrame,
+        datasetB: DataFrame,
+        numNearestNeighbors: int,
+        distCol: str = "distCol",
+    ) -> DataFrame:
+        """
+        This function returns the k exact nearest neighbors (knn) in datasetA regarding each query vector in datasetB.
+        The purpose is to match pyspark LSH approxSimilarityJoin(datasetA, datasetB, threshold, distCol). Note this function
+        is directional that treats datasetA as the item dataframe and datasetB as the query dataframe. This is because
+        knn-based search requires the role of query vectors, which is different from threshold-based search where datasetA and
+        datasetB are interchangeable.
+
+        Parameters
+        ----------
+        datasetA: pyspark.sql.DataFrame
+            the item_df dataframe from which the k nearest neighbors will be retrieved. Each row represents an item vector.
+
+        datasetB: pyspark.sql.DataFrame
+            the query_df dataframe. Each row represents a query vector.
+
+        numNearestNeighbors: int
+            the number of exact nearest neighbors to be returned for each query.
+
+        distCol: str
+            the name of the distance column
+
+        Returns
+        -------
+        knnjoin_df: pyspark.sql.DataFrame
+            the result dataframe that has three columns (datasetA, datasetB, distCol).
+            datasetA column is of struct type that concatenates all the columns of input dataframe datasetA.
+            Similarly, datasetB column is of struct type that concatenates all the columns of input dataframe datasetB.
+            distCol is the distance column. A row in knnjoin_df is in the format (v1, v2, dist(v1, v2)),
+            where item_vector v1 is one of the k nearest neighbors of query_vector v2 and their distance is dist(v1, v2).
```

Comment:
Again, this ordering seems backward to me.  Can you explain rationale?

Reply:
See reply in the next comment.